### ISSUE TITLE
alias rewriting can't handle quote in path

### ISSUE THREAD

**Issue by wmertens** [pnp/pnpjs#2222]
### Category
- Bug

### Version

Please specify what version of the library you are using: [        v3 ]

Please specify what version(s) of SharePoint you are targeting: [online         ]

### Expected / Desired Behavior / Question

The `!@x::` prefix always results in a correct string

### Observed Behavior

When trying to alias the path `/sites/Testing/Shared Documents/file\u015bTest-'2\uff01/test.pptx` as embedded by `web.getFileByUrl()`, it only extracts the path up to `'`. The problem is that the alias is extracted from the URL without encoding/decoding.

I think calls like Web.getFileByUrl should convert `'` to `''` and alias handling should reverse that again. URI encoding is also an option but that's a lot more likely to impact existing code.

```
Verbose Rewriting aliased parameter from match '!@p1::%2Fsites%2FTesting%2FShared%20Documents%2Ffile%C5%9BTest-' to label: @p1 value: %2Fsites%2FTesting%2FShared%20Documents%2Ffile%C5%9BTest-
[...]
Info [728293a4-00a3-41f5-a162-630d24a9748d] Url: https://stratokitio.sharepoint.com/sites/Testing/_api/web/getFileByUrl(@p1'2%EF%BC%81%2Ftest.pptx')/startUpload(uploadId=guid'14333e82-8540-4000-a800-6dec1768ce0a')?@p1='%2Fsites%2FTesting%2FShared%20Documents%2Ffile%C5%9BTest-'
```

**Comment by juliemturner** [comment]
Duplicate of #2220

**Comment by wmertens** [comment]
Ah - I opened it as a separate issue because this is about alias handling and the other issue's code doesn't even use aliases, but if all path and name references throughout the code were aliased it would indeed solve both issues.

A thought occurs, how about a special alias name like @@@ that would be replaced by a per-call incrementing name @_1 etc so we don't need to worry about clashing alias names?

**Comment by wmertens** [comment]
I see two problems in https://github.com/pnp/pnpjs/blob/c0b3cb0e6b8f7b9f0272f238ea04f9784de5457f/packages/sp/spqueryable.ts#L81-L97 : 
- the regex should have `(!?')` at the end to negatively lookahead a second quote
- the parameters should be URI encoded before embedding in the URL

**Comment by juliemturner** [comment]
The fix for this issue was to use a lookback regex, unfortunately unbeknownst to us as the time this broke the Safari (iOS) browser. We've reversed the fix (3.4.1 release) so that we can unblock those targeting that browser.
